[JSC] Array#concat should first create a butterfly before result array

Constellation opened the pull request:
#### 172ece169a8045a8239dd635d8d0000b3032708c
<pre>
[JSC] Array#concat should first create a butterfly before result array
<a href="https://bugs.webkit.org/show_bug.cgi?id=289391">https://bugs.webkit.org/show_bug.cgi?id=289391</a>
<a href="https://rdar.apple.com/146520816">rdar://146520816</a>

Reviewed by Sosuke Suzuki.

We should just first allocate a butterfly, copying content, and then
create an array from that since copying does not cause any GC and
butterfly&apos;s content is coming from the other arrays which is under GC
management. So in this case, we do not need to use GCSafe ops since
these butterflies are not scanned until it gets set to an array, which
will be done after setting everything.

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::concatAppendArray):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::tryCloneArrayFromFast):
* Source/JavaScriptCore/runtime/JSArray.h:
(JSC::copyArrayElements):

Canonical link: <a href="https://commits.webkit.org/291828@main">https://commits.webkit.org/291828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a5f45ed6e3486116316194b9821e5eecc9b5eb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94145 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99156 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44672 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22162 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71812 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29151 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97147 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10399 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84989 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52153 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2677 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43990 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86855 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80324 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2767 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101198 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92811 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21197 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80816 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21449 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81005 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80198 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24742 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14359 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15097 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21181 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26360 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115461 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20868 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24328 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22609 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1723,31 +1723,38 @@ static JSArray* concatAppendArray(JSGlobalObject* globalObject, VM& vm, JSArray*
 
-    ASSERT(result->butterfly()->publicLength() == resultSize);
-    return result;
+    return JSArray::createWithButterfly(vm, nullptr, resultStructure, butterfly);
```
Should we set up to `vectorLength` with some valid values before setting it to `createWithButterfly` like https://github.com/WebKit/WebKit/commit/9158c52898ef7f10c47c884c12c67de5ee47d711 ?

Constellation replied:
Right. Done.

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -1961,40 +1961,48 @@ JSArray* tryCloneArrayFromFast(JSGlobalObject* globalObject, JSValue arrayValue)
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+    return JSArray::createWithButterfly(vm, nullptr, resultStructure, resultButterfly);
```
ditto

Constellation replied:
Done.

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSArray.h:
```diff
@@ -349,7 +355,7 @@ ALWAYS_INLINE void copyArrayElements(T* buffer, unsigned offset, U* source, unsi
         if constexpr (fillMode == ArrayFillMode::Empty) {
-            if constexpr (std::is_same_v<T, double>)
+            if constexpr (std::is_same_v<T, double> || needsGCSafeOps == NeedsGCSafeOps::No)
                 memcpy(buffer + offset, source, sizeof(double) * sourceSize);
```
I think the contents of buffer now can be non-double value. But it is okay to leave it as `sizeof(double)`?
( In my knowledge, JSValue and double are the same size, but just in case. )

Constellation replied:
We can add `static_assert` around size.

sosukesuzuki reviewed (commented):
Maybe we should address safer cpp style errors?

Constellation committed:
[JSC] Array#concat should first create a butterfly before result array
https://bugs.webkit.org/show_bug.cgi?id=289391
rdar://146520816

Reviewed by Sosuke Suzuki.

We should just first allocate a butterfly, copying content, and then
create an array from that since copying does not cause any GC and
butterfly's content is coming from the other arrays which is under GC
management. So in this case, we do not need to use GCSafe ops since
these butterflies are not scanned until it gets set to an array, which
will be done after setting everything.

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::concatAppendArray):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::tryCloneArrayFromFast):
* Source/JavaScriptCore/runtime/JSArray.h:
(JSC::copyArrayElements):

Canonical link: https://commits.webkit.org/291828@main
